### PR TITLE
FAST multitenant bootstrap and resource management, rename org-level FAST stages

### LINKED ISSUE #1050: Create a script to link files to fast stages
via @agutta

### PR DESCRIPTION
**Pull request by ludoo**
This PR

- implements single tenant bootstrap and resource management stages
- renames all FAST stages
- implements optional tag key creation in the organization module
- adds a script to output links needed for each stage (fixes #1050)

There are a couple of features which will be nice to add to tenant bootstrap but which can wait for a subsequent PR, they are noted in its README. The security stage will also need to be changed accept a policy at the folder level, as this is one of the use cases which make sense.

### CODE REVIEW

**Comment by ludoo** [comment]
Marked as ready, we need to cut a release before merging this to save the "last known good state" of FAST stages before renaming.

File: fast/stages-multitenant/0-bootstrap-tenant/organization.tf [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    },
+    var.billing_account.is_org_level ? {
+      "roles/billing.admin" = [
+        "group:${local.groups.gcp-admins}",
+        module.automation-tf-resman-sa.iam_email
+      ]
+      "roles/billing.costsManager" = ["group:${local.groups.gcp-admins}"]
+    } : {}
+  )
+  tags = {
```
Comment (juliocc) [inline fast/stages-multitenant/0-bootstrap-tenant/organization.tf:43]: I might be missing something but I think you're ignoring all the other tags in `var.tag_values` and `var.tag_keys`

**Review by juliocc — COMMENTED** [review]
I'll do this one in parts. Here's a few comments/questions on the tenant bootstrap.

File: fast/stages-multitenant/0-bootstrap-tenant/billing.tf [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/**
```
[the remark this answers is not in the document]
Comment (ludoo) [reply fast/stages-multitenant/0-bootstrap-tenant/billing.tf:1]: the org-level bootstrap creates billing IAM regardless, it just changes where (org or ba) depending on that variable; conditionally applying billing IAM means having service accounts that conditionally create projects, which would not work :)

File: fast/stages-multitenant/0-bootstrap-tenant/variables.tf [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+# defaults for variables marked with global tfdoc annotations, can be set via
+# the tfvars file generated in stage 00 and stored in its outputs
+
+variable "automation" {
+  # tfdoc:variable:source 0-0-bootstrap
+  description = "Automation resources created by the bootstrap stage."
```
[the remark this answers is not in the document]
Comment (ludoo) [reply fast/stages-multitenant/0-bootstrap-tenant/variables.tf]: Done.

File: fast/stages-multitenant/0-bootstrap-tenant/variables.tf [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    teams           = bool
+  })
+  default = {
+    data_platform   = true
+    gke             = true
+    project_factory = true
+    sandbox         = true
+    teams           = true
+  }
+  # nullable = false
```
[the remark this answers is not in the document]
Comment (ludoo) [reply fast/stages-multitenant/0-bootstrap-tenant/variables.tf]: yep and changed to optional

File: fast/stages-multitenant/0-bootstrap-tenant/main.tf [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    ? data.google_client_openid_userinfo.resman-sa.0.email
+    : var.test_principal
+  )
+}
+
+data "google_client_openid_userinfo" "resman-sa" {
+  count = var.test_principal == null ? 1 : 0
+}
+
+module "tenant-folder" {
```
[the remark this answers is not in the document]
Comment (ludoo) [reply fast/stages-multitenant/0-bootstrap-tenant/main.tf:43]: I kind of prefer having them in main, like we do in blueprints for projects, etc. If you feel strongly about this no probs, it's a 10s change we can of course do it.

File: fast/stages-multitenant/0-bootstrap-tenant/main.tf [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  source = "../../../modules/folder"
+  parent = "organizations/${var.organization.id}"
+  name   = var.tenant_config.descriptive_name
+  tag_bindings = {
+    tenant = try(
+      module.organization.tag_values["${var.tag_names.tenant}/${var.tenant_config.short_name}"].id,
+      null
+    )
+  }
+}
```
[the remark this answers is not in the document]
Comment (ludoo) [reply fast/stages-multitenant/0-bootstrap-tenant/main.tf:53]: yowz!

[the remark this answers is not in the document]
Comment (ludoo) [reply fast/stages-multitenant/0-bootstrap-tenant/main.tf:61]: done! great catch

File: fast/stages-multitenant/0-bootstrap-tenant/organization.tf (hunk 1 above)
Comment (ludoo) [reply fast/stages-multitenant/0-bootstrap-tenant/organization.tf:43]: yep we are, those are produced for consumption by everything that comes after org resman, here we only need tenant

**Review by juliocc — COMMENTED** [review]
IAM.md files are out of date

File: fast/stages-multitenant/1-resman-tenant/variables.tf [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+  # tfdoc:variable:source 0-bootstrap
+  description = "Organization details."
+  type = object({
+    domain      = string
+    id          = number
+    customer_id = string
+  })
+}
+
+variable "organization_policy_data_path" {
```
[the remark this answers is not in the document]
Comment (ludoo) [reply fast/stages-multitenant/1-resman-tenant/variables.tf:216]: we will align the org-level stage to this convention after merging, as discussed

File: fast/stages-multitenant/1-resman-tenant/branch-project-factory.tf [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+# tfdoc:file:description Project factory stage resources.
+
```
[the remark this answers is not in the document]
Comment (ludoo) [reply fast/stages-multitenant/1-resman-tenant/branch-project-factory.tf:18]: added

File: fast/stages-multitenant/1-resman-tenant/branch-data-platform.tf [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    "roles/owner"                          = [local.automation_sas_iam.dp-prod]
+    "roles/logging.admin"                  = [local.automation_sas_iam.dp-prod]
+    "roles/resourcemanager.folderAdmin"    = [local.automation_sas_iam.dp-prod]
+    "roles/resourcemanager.projectCreator" = [local.automation_sas_iam.dp-prod]
+  }
+  tag_bindings = {
+    context = var.tags.values["${var.tags.names.environment}/production"]
+  }
+}
+
```
[the remark this answers is not in the document]
Comment (ludoo) [reply fast/stages-multitenant/1-resman-tenant/branch-data-platform.tf:70]: added

File: fast/stages-multitenant/1-resman-tenant/main.tf [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    )
+  }
+  cicd_workflow_var_files = {
+    stage_2 = [
+      "0-bootstrap-tenant.auto.tfvars.json",
+    ]
+    stage_3 = [
+      "0-bootstrap-tenant.auto.tfvars.json",
+      "2-networking.auto.tfvars.json",
+      "2-security.auto.tfvars.json"
```
[the remark this answers is not in the document]
Comment (ludoo) [reply fast/stages-multitenant/1-resman-tenant/main.tf:63]: they use the same stages as the org, so I would omit; they will be created in the tenant folder so tenant is there

File: fast/stages-multitenant/1-resman-tenant/variables.tf [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+      issuer_uri       = string
+      name             = string
+      principal_tpl    = string
+      principalset_tpl = string
+    }))
+    service_accounts = object({
+      networking = string
+      resman     = string
+      security   = string
+      dp-dev     = optional(string)
```
[the remark this answers is not in the document]
Comment (ludoo) [reply fast/stages-multitenant/1-resman-tenant/variables.tf:39]: they all depend on fast features, so the only sistuation where that might happen is someone forcing fast features here, which they should not do

**Pull request merged**